### PR TITLE
Search for assetitem

### PR DESCRIPTION
**Pull request by alina612**
This pr also includes #75. 
Only the following files were changed:

src/TuDa.CIMS.Web/Services/IAssetItemApi.cs
src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs
src/TuDa.CIMS.Api/Interfaces/IAssetItemRepository.cs
src/TuDa.CIMS.Api/Interfaces/IAssetItemService.cs
src/TuDa.CIMS.Api/Services/AssetItemService.cs
src/TuDa.CIMS.Api/Controllers/AssetItemController.cs
src/TuDa.CIMS.Web/Components/ShoppingSearch.razor
src/TuDa.CIMS.Web/Components/ShoppingSearch.razor.cs

The search bar needs to be placed in the correct position in the ShoppingCartPage ticket.
search

### CODE REVIEW

File: src/TuDa.CIMS.Api/Controllers/AssetItemController.cs [hunk 1]
```diff
@@ -22,9 +23,9 @@ public AssetItemController(IAssetItemService assetItemService)
     /// </summary>
     /// <returns> a 200 OK response if the operation is successfully and a 400 BadRequest response if any error occurs </returns>
     [HttpGet]
-    public async Task<IActionResult> GetAllAsync()
+    public async Task<IActionResult> GetAllAsync([Query] string? nameOrCas)
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Controllers/AssetItemController.cs:26]: 🔧 Unfortunately, that's the wrong attribute. This is from `Refit,` but we need the `FromQuery` from Asp.Net
```suggestion
    public async Task<IActionResult> GetAllAsync([FromQuery] string? nameOrCas)
```

File: src/TuDa.CIMS.Api/Interfaces/IAssetItemService.cs [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
+﻿using Refit;
+using TuDa.CIMS.Shared.Dtos;
 using TuDa.CIMS.Shared.Entities;
 
 namespace TuDa.CIMS.Api.Interfaces;
 
 public interface IAssetItemService
 {
-    Task<ErrorOr<IEnumerable<AssetItem>>> GetAllAsync();
+    Task<ErrorOr<IEnumerable<AssetItem>>> GetAllAsync([Query] string? nameOrCas);
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Interfaces/IAssetItemService.cs:9]: 🔧 Attribute does nothing here
```suggestion
    Task<ErrorOr<IEnumerable<AssetItem>>> GetAllAsync(string? nameOrCas);
```

File: src/TuDa.CIMS.Api/Services/AssetItemService.cs [hunk 3]
```diff
@@ -16,21 +17,28 @@ public AssetItemService(IAssetItemRepository assetItemRepository)
 
     /// <summary>
     /// Return an an <see cref="ErrorOr{T}"/> that either contains an error message if an error occurs,
-    /// or the result of the <see cref="GetAllAsync"/> functionality if successful
+    /// or the result of the <see cref="GetAllAsync"/> functionality if successful or if nameOrCas is set the <see cref="SearchAsync"/> functionality.
     /// </summary>
 
-    public async Task<ErrorOr<IEnumerable<AssetItem>>> GetAllAsync()
+    public async Task<ErrorOr<IEnumerable<AssetItem>>> GetAllAsync([Query] string? nameOrCas)
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Services/AssetItemService.cs:23]: 🔧 Attribute does nothing here

File: src/TuDa.CIMS.Web/Components/ShoppingSearch.razor [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿@page "/SearchBar"
+<MudGrid Justify="Justify.Center">
+    <MudItem xs="12" sm="8" md="8">
+        <MudAutocomplete
+            @bind-Value="selectedItem":
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Web/Components/ShoppingSearch.razor:5]: 🔧 Remove this here
```suggestion
            @bind-Value="selectedItem"
```

File: src/TuDa.CIMS.Web/Components/ShoppingSearch.razor.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace TuDa.CIMS.Web.Components;
+
+/// <summary>
+///Represents a component that allows users to search for and select an <see cref="AssetItem"/> on the ShoppingCart Page.
+/// </summary>
+public partial class ShoppingSearch : ComponentBase
+{
+    private readonly IAssetItemApi _assetItemApi;
+    private AssetItem selectedItem;
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Web/Components/ShoppingSearch.razor.cs:13]: 🔧 Can be safely removed

**Review by Sc3l3t0n — CHANGES_REQUESTED** [review]

File: src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs [hunk 6]
```diff
@@ -137,4 +137,27 @@ public async Task<ErrorOr<Deleted>> RemoveAsync(Guid id)
...
+    ///Returns a list of matching AssetItem based on the provided name or CAS number.
+    /// </summary>
+    /// <param name="nameOrCas"></param>
+    public async Task<IEnumerable<AssetItem>> SearchAsync(string nameOrCas)
+    {
+        IQueryable<AssetItem> query;
+
+        bool isCas = nameOrCas.All(c => char.IsDigit(c) || c == '-');
+
+        //TODO fix include hazards
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs:151]: 🔧 The changes should be merged and you can include the Hazards again

File: src/TuDa.CIMS.Web/Components/ShoppingSearch.razor.cs [hunk 7]
```diff
@@ -22,6 +24,13 @@ public ShoppingSearch(IAssetItemApi api)
         _assetItemApi = api;
     }
 
+    /// Invoke to clear text
+    private async Task SelectAssetItem(AssetItem item)
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Web/Components/ShoppingSearch.razor.cs:28]: ⛏️  Would rename it to `AssetItemSelectedInternal`

File: src/TuDa.CIMS.Web/Components/ShoppingSearch.razor [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+            Clearable="@true"
+            AdornmentIcon="@Icons.Material.Filled.Search"
+            AdornmentColor="@Color.Primary"
+            Label="Items"
+            SearchFunc="@Search"
+            ToStringFunc="@ToString"
+            Variant="Variant.Filled"
+            Dense=true
+            ShowProgressIndicator="true">
+        </MudAutocomplete>
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Web/Components/ShoppingSearch.razor:14]: 🔧 Add back in the `ValueChanged`. I found out what were wrong, you need to add the Type back in and then he takes a normal function as parameter.
```suggestion
        <MudAutocomplete
            T="AssetItem"
            @ref="_autocomplete"
            Clearable="@true"
            AdornmentIcon="@Icons.Material.Filled.Search"
            AdornmentColor="@Color.Primary"
            Label="Items"
            SearchFunc="@Search"
            ToStringFunc="@ToString"
            ValueChanged="@AssetItemSelectedInternal"
            Variant="Variant.Filled"
            Dense=true
            ShowProgressIndicator="true">
        </MudAutocomplete>
```

File: src/TuDa.CIMS.Web/Components/ShoppingSearch.razor.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace TuDa.CIMS.Web.Components;
+
+/// <summary>
+///Represents a component that allows users to search for and select an <see cref="AssetItem"/> on the ShoppingCart Page.
+/// </summary>
+public partial class ShoppingSearch : ComponentBase
+{
+    private readonly IAssetItemApi _assetItemApi;
+    private MudAutocomplete<AssetItem> _autocomplete;
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Web/Components/ShoppingSearch.razor.cs:14]: 🔧 Add this to suppress the warning
```suggestion
    private MudAutocomplete<AssetItem> _autocomplete = null!; // Is set by blazor component
```

File: src/TuDa.CIMS.Web/Components/ShoppingSearch.razor.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+        return await _assetItemApi.GetAllAsync(nameOrCas).Match(value => value, err => []);
+    }
+
+    private static string ToString(AssetItem item) =>
+        item switch
+        {
+            Substance substance => $"{substance.Name} ({substance.Cas})",
+            _ => item.Name,
+        };
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Web/Components/ShoppingSearch.razor.cs:49]: 🔧 I got a null dereference error, so maybe add a null case. 
```suggestion
    private static string ToString(AssetItem item) =>
        item switch
        {
            Substance substance => $"{substance.Name} ({substance.Cas})",
            null => "",
            _ => item.Name,
        };
```

**Review by Sc3l3t0n — CHANGES_REQUESTED** [review]

**Review by Sc3l3t0n — APPROVED** [review]

**Review by anngy12 — APPROVED** [review]

**Pull request merged**